components/server/src/authorization/relationship-updater.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { ProjectDB, TeamDB, UserDB } from "@gitpod/gitpod-db/lib";
+import { AdditionalUserData, Organization, User } from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable } from "inversify";
+import { RedisMutex } from "../redis/mutex";
+import { Authorizer } from "./authorizer";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+
+@injectable()
+export class RelationshipUpdater {
+    public readonly version = 1;
+
+    constructor(
+        @inject(RedisMutex) private readonly mutex: RedisMutex,
+        @inject(UserDB) private readonly userDB: UserDB,
+        @inject(TeamDB) private readonly orgDB: TeamDB,
+        @inject(ProjectDB) private readonly projectDB: ProjectDB,
+        @inject(Authorizer) private readonly authorizer: Authorizer,
+    ) {}
+
+    /**
+     * Updates all relationships for a user according to the current state of the database.
+     * @param user
+     * @returns
+     */
+    public async migrate(user: User): Promise<User> {
+        if (user.additionalData?.fgaRelationshipsVersion === this.version) {
```

Comment:
What are the semantics of `fgaRelationshipsVersion`? Is a user "stepping" through versions, or why is this a `<=`? :thinking:

Reply:
Because the migrator is idempotent we can support downgrades as well in case we want to rollback a change or just revert a commit that has a version increment in it.

Reply:
Gotcha, so:
  - it will only ever be this file that set that field
  - changes/versions will be handled via Git 
  - it's intended to always be idempotent

Can this be part of the class/method docs? :orange_heart: